vcd/resource_vcd_edgegateway.go

```diff
@@ -0,0 +1,201 @@
+				Description: "If advanced networking enabled, also enable distributed routing",
+			},
+		},
+	}
+}
+
+// Creates a new edge gateway from a resource definition
+func resourceVcdEdgeGatewayCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[TRACE] edge gateway creation initiated")
+
+	vcdClient := meta.(*VCDClient)
+
+	rawExternalNetworks := d.Get("external_networks").([]interface{})
+	var externalNetworks []string
+	for _, en := range rawExternalNetworks {
+		externalNetworks = append(externalNetworks, en.(string))
+	}
+
+	var gwCreation = govcd.EdgeGatewayCreation{
+		ExternalNetworks:          externalNetworks,
+		Name:                      d.Get("name").(string),
+		OrgName:                   d.Get("org").(string),
+		VdcName:                   d.Get("vdc").(string),
+		Description:               d.Get("description").(string),
+		BackingConfiguration:      d.Get("backing_config").(string),
+		AdvancedNetworkingEnabled: d.Get("advanced_networking").(bool),
+		DefaultGateway:            d.Get("default_gateway").(string),
+		DistributedRoutingEnabled: d.Get("distributed_routing").(bool),
+		HAEnabled:                 d.Get("ha_enabled").(bool),
+	}
+
+	edge, err := govcd.CreateEdgeGateway(vcdClient.VCDClient, gwCreation)
+	if err != nil {
+		log.Printf("[DEBUG] Error creating edge gateway: %#v", err)
+		return fmt.Errorf("error creating edge gateway: %#v", err)
+	}
+
+	d.SetId(d.Get("name").(string))
+	log.Printf("[TRACE] edge gateway created: %#v", edge.EdgeGateway.Name)
+	return resourceVcdEdgeGatewayRead(d, meta)
+}
+
+func setEdgeGatewayValues(d *schema.ResourceData, egw govcd.EdgeGateway) error {
+
+	d.SetId(egw.EdgeGateway.HREF)
+	err := d.Set("name", egw.EdgeGateway.Name)
+	if err != nil {
+		return err
+	}
+	err = d.Set("description", egw.EdgeGateway.Description)
+	if err != nil {
+		return err
+	}
+	err = d.Set("backing_config", egw.EdgeGateway.Configuration.GatewayBackingConfig)
+	if err != nil {
+		return err
+	}
+	var networks []string
+	for _, net := range egw.EdgeGateway.Configuration.GatewayInterfaces.GatewayInterface {
+		networks = append(networks, net.Network.Name)
+	}
+
+	err = d.Set("external_networks", networks)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Fetches information about an existing edge gateway for a data definition
+func resourceVcdEdgeGatewayRead(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[TRACE] edge gateway read initiated")
+
+	vcdClient := meta.(*VCDClient)
+
+	edgeGateway, err := vcdClient.GetEdgeGatewayFromResource(d, "name")
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("error fetching edge gateway details %#v", err)
+	}
+
+	err = setEdgeGatewayValues(d, edgeGateway)
+
+	log.Printf("[TRACE] edge gateway read completed: %#v", edgeGateway.EdgeGateway)
+	return err
+}
+
+// Deletes a edge gateway, optionally removing all objects in it as well
+func resourceVcdEdgeGatewayDelete(d *schema.ResourceData, meta interface{}) error {
```

Comment:
lock is needed to be safe

Reply:
Added

Reply:
Now it locks everything, we need only lock this edge gtw -> recommend create function in config.go `lockEdgeGtw()` which will use `name` for `key`

Reply:
Everything is fine, my bad